File: dart-jaguar/lib/api/messages_api.dart
import 'package:jaguar_retrofit/annotations/annotations.dart';
import 'package:jaguar_retrofit/jaguar_retrofit.dart';
import 'package:jaguar_serializer/jaguar_serializer.dart';
import 'package:jaguar_mimetype/jaguar_mimetype.dart';
import 'dart:async';

import 'package:openapi/model/json_success_base.dart';
import 'package:openapi/model/json_success.dart';
import 'package:openapi/model/one_ofstringinteger.dart';
import 'package:openapi/model/one_ofobjectobject.dart';
import 'package:openapi/model/invalid_message_error.dart';
import 'package:openapi/model/coded_error.dart';

part 'messages_api.jretro.dart';

@GenApiClient()
class MessagesApi extends ApiClient with _$MessagesApiClient {
    final Route base;
    final Map<String, CodecRepo> converters;
    final Duration timeout;

    MessagesApi({this.base, this.converters, this.timeout = const Duration(minutes: 2)});

    /// Add an emoji reaction
    ///
    /// Add an [emoji reaction](/help/emoji-reactions) to a message.  &#x60;POST {{ api_url }}/v1/messages/{message_id}/reactions&#x60; 
    @PostReq(path: "/messages/:message_id/reactions")
    Future<JsonSuccess> addReaction(
            @PathParam("message_id") int messageId
        ,
            @QueryParam("emoji_name") String emojiName, 
        
            @QueryParam("emoji_code") String emojiCode, 
        
            @QueryParam("reaction_type") String reactionType
        ) {
        return super.addReaction(
        messageId
        ,
        emojiName, 
        
        emojiCode, 
        
        reactionType

        ).timeout(timeout);
    }

    /// Check if messages match a narrow
    ///
    /// Check whether a set of messages match a [narrow](/api/construct-narrow).  &#x60;GET {{ api_url }}/v1/messages/matches_narrow&#x60;  For many common narrows (E.g. a topic), clients can write an efficient client-side check to determine whether a newly arrived message belongs in the view.  This endpoint is designed to allow clients to handle more complex narrows for which the client does not (or in the case of full-text search, cannot) implement this check.  The format of the &#x60;match_subject&#x60; and &#x60;match_content&#x60; objects is designed to match those of &#x60;GET /messages&#x60;, so that a client can splice these fields into a &#x60;message&#x60; object received from &#x60;GET /events&#x60; and end up with an extended message object identical to how a &#x60;GET /messages&#x60; for the current narrow would have returned the message. 
    @GetReq(path: "/messages/matches_narrow")
    Future<Object> checkMessagesMatchNarrow(
        
            @QueryParam("msg_ids") List<int> msgIds, 
        
            @QueryParam("narrow") List<Object> narrow
        ) {
        return super.checkMessagesMatchNarrow(
        
        msgIds, 
        
        narrow

        ).timeout(timeout);
    }

    /// Delete a message
    ///
    /// Permanently delete a message.  &#x60;DELETE {{ api_url }}/v1/messages/{msg_id}&#x60;  This API corresponds to the [delete a message completely][delete-completely] feature documented in the Zulip Help Center.  [delete-completely]: /help/edit-or-delete-a-message#delete-a-message-completely 
    @DeleteReq(path: "/messages/:message_id")
    Future<JsonSuccess> deleteMessage(
            @PathParam("message_id") int messageId
        ) {
        return super.deleteMessage(
        messageId

        ).timeout(timeout);
    }

    /// Get public temporary URL
    ///
    /// Get a temporary URL for access to the file that doesn&#39;t require authentication. 
    @GetReq(path: "/user_uploads/:realm_id_str/:filename")
    Future<JsonSuccessBase> getFileTemporaryUrl(
            @PathParam("realm_id_str") int realmIdStr, 
            @PathParam("filename") String filename
        ) {
        return super.getFileTemporaryUrl(
        realmIdStr, 
        filename

        ).timeout(timeout);
    }

    /// Get a message&#39;s edit history
    ///
    /// Fetch the message edit history of a previously edited message.  &#x60;GET {{ api_url }}/v1/messages/{message_id}/history&#x60;  Note that edit history may be disabled in some organizations; see the [Zulip Help Center documentation on editing messages][edit-settings].  [edit-settings]: /help/view-a-messages-edit-history 
    @GetReq(path: "/messages/:message_id/history")
    Future<JsonSuccessBase> getMessageHistory(
            @PathParam("message_id") int messageId
        ) {
        return super.getMessageHistory(
        messageId

        ).timeout(timeout);
    }

    /// Get messages
    ///
    /// Fetch message history from a Zulip server.  &#x60;GET {{ api_url }}/v1/messages&#x60;  This &#x60;GET /api/v1/messages&#x60; endpoint is the primary way to fetch message history from a Zulip server.  It is useful both for Zulip clients (e.g. the web, desktop, mobile, and terminal clients) as well as bots, API clients, backup scripts, etc.  By specifying a [narrow filter](/api/construct-narrow), you can use this endpoint to fetch the messages matching any search query that is supported by Zulip&#39;s powerful full-text search backend.  When a narrow is not specified, it can be used to fetch a user&#39;s message history. (We recommend paginating to 1000 messages at a time.)  In either case, you specify an &#x60;anchor&#x60; message (or ask the server to calculate the first unread message for you and use that as the anchor), as well as a number of messages before and after the anchor message.  The server returns those messages, sorted by message ID, as well as some metadata that makes it easy for a client to determine whether there are more messages matching the query that were not returned due to the &#x60;num_before&#x60; and &#x60;num_after&#x60; limits.  We recommend using &#x60;num_before &lt;&#x3D; 1000&#x60; and &#x60;num_after &lt;&#x3D; 1000&#x60; to avoid generating very large HTTP responses. A maximum of 5000 messages can be obtained per request; attempting to exceed this will result in an error. 
    @GetReq(path: "/messages")
    Future<JsonSuccessBase> getMessages(
        
            @QueryParam("anchor") OneOfstringinteger anchor, 
        
            @QueryParam("num_before") int numBefore, 
        
            @QueryParam("num_after") int numAfter, 
        
            @QueryParam("narrow") List<Object> narrow, 
        
            @QueryParam("client_gravatar") bool clientGravatar, 
        
            @QueryParam("apply_markdown") bool applyMarkdown, 
        
            @QueryParam("use_first_unread_anchor") bool useFirstUnreadAnchor
        ) {
        return super.getMessages(
        
        anchor, 
        
        numBefore, 
        
        numAfter, 
        
        narrow, 
        
        clientGravatar, 
        
        applyMarkdown, 
        
        useFirstUnreadAnchor

        ).timeout(timeout);
    }

    /// Get a message&#39;s raw Markdown
    ///
    /// Get the raw content of a message.  &#x60;GET {{ api_url }}/v1/messages/{msg_id}&#x60;  This is a rarely-used endpoint relevant for clients that primarily work with HTML-rendered messages but might need to occasionally fetch the message&#39;s raw Markdown (e.g. for pre-filling a message-editing UI). 
    @GetReq(path: "/messages/:message_id")
    Future<JsonSuccessBase> getRawMessage(
            @PathParam("message_id") int messageId
        ) {
        return super.getRawMessage(
        messageId

        ).timeout(timeout);
    }

    /// Mark all messages as read
    ///
    /// Marks all of the current user&#39;s unread messages as read.  &#x60;POST {{ api_url }}/v1/mark_all_as_read&#x60; 
    @PostReq(path: "/mark_all_as_read")
    Future<JsonSuccess> markAllAsRead(
        ) {
        return super.markAllAsRead(

        ).timeout(timeout);
    }

    /// Mark messages in a stream as read
    ///
    /// Mark all the unread messages in a stream as read. 
    @PostReq(path: "/mark_stream_as_read")
    Future<JsonSuccess> markStreamAsRead(
        
            @QueryParam("stream_id") int streamId
        ) {
        return super.markStreamAsRead(
        
        streamId

        ).timeout(timeout);
    }

    /// Mark messages in a topic as read
    ///
    /// Mark all the unread messages in a topic as read. 
    @PostReq(path: "/mark_topic_as_read")
    Future<JsonSuccess> markTopicAsRead(
        
            @QueryParam("stream_id") int streamId, 
        
            @QueryParam("topic_name") String topicName
        ) {
        return super.markTopicAsRead(
        
        streamId, 
        
        topicName

        ).timeout(timeout);
    }

    /// Remove an emoji reaction
    ///
    /// Remove an [emoji reaction](/help/emoji-reactions) from a message.  &#x60;DELETE {{ api_url }}/v1/messages/{message_id}/reactions&#x60; 
    @DeleteReq(path: "/messages/:message_id/reactions")
    Future<JsonSuccess> removeReaction(
            @PathParam("message_id") int messageId
        ,
            @QueryParam("emoji_name") String emojiName, 
        
            @QueryParam("emoji_code") String emojiCode, 
        
            @QueryParam("reaction_type") String reactionType
        ) {
        return super.removeReaction(
        messageId
        ,
        emojiName, 
        
        emojiCode, 
        
        reactionType

        ).timeout(timeout);
    }

    /// Render message
    ///
    /// Render a message to HTML.  &#x60;POST {{ api_url }}/v1/messages/render&#x60; 
    @PostReq(path: "/messages/render")
    Future<JsonSuccessBase> renderMessage(
        
            @QueryParam("content") String content
        ) {
        return super.renderMessage(
        
        content

        ).timeout(timeout);
    }

    /// Send a message
    ///
    /// Send a stream or a private message.  &#x60;POST {{ api_url }}/v1/messages&#x60; 
    @PostReq(path: "/messages")
    Future<JsonSuccessBase> sendMessage(
        
            @QueryParam("type") String type, 
        
            @QueryParam("to") List<int> to, 
        
            @QueryParam("content") String content, 
        
            @QueryParam("topic") String topic, 
        
            @QueryParam("queue_id") String queueId, 
        
            @QueryParam("local_id") String localId
        ) {
        return super.sendMessage(
        
        type, 
        
        to, 
        
        content, 
        
        topic, 
        
        queueId, 
        
        localId

        ).timeout(timeout);
    }

    /// Edit a message
    ///
    /// Edit/update the content or topic of a message.  &#x60;PATCH {{ api_url }}/v1/messages/{msg_id}&#x60;  &#x60;{msg_id}&#x60; in the above URL should be replaced with the ID of the message you wish you update.  You can [resolve topics](/help/resolve-a-topic) by editing the topic to &#x60;??? {original_topic}&#x60;.  **Note**: See [configuring message editing][config-message-editing] for detailed documentation on when users are allowed to edit topics.  [config-message-editing]: /help/configure-message-editing-and-deletion 
    @PatchReq(path: "/messages/:message_id")
    Future<JsonSuccess> updateMessage(
            @PathParam("message_id") int messageId
        ,
            @QueryParam("topic") String topic, 
        
            @QueryParam("propagate_mode") String propagateMode, 
        
            @QueryParam("send_notification_to_old_thread") bool sendNotificationToOldThread, 
        
            @QueryParam("send_notification_to_new_thread") bool sendNotificationToNewThread, 
        
            @QueryParam("content") String content, 
        
            @QueryParam("stream_id") int streamId
        ) {
        return super.updateMessage(
        messageId
        ,
        topic, 
        
        propagateMode, 
        
        sendNotificationToOldThread, 
        
        sendNotificationToNewThread, 
        
        content, 
        
        streamId

        ).timeout(timeout);
    }

    /// Update personal message flags
    ///
    /// Add or remove personal message flags like &#x60;read&#x60; and &#x60;starred&#x60; on a collection of message IDs.  &#x60;POST {{ api_url }}/v1/messages/flags&#x60;  For updating the &#x60;read&#x60; flag on common collections of messages, see also the [special endpoints for marking message as read in bulk](/api/mark-all-as-read). 
    @PostReq(path: "/messages/flags")
    Future<JsonSuccessBase> updateMessageFlags(
        
            @QueryParam("messages") List<int> messages, 
        
            @QueryParam("op") String op, 
        
            @QueryParam("flag") String flag
        ) {
        return super.updateMessageFlags(
        
        messages, 
        
        op, 
        
        flag

        ).timeout(timeout);
    }

    /// Upload a file
    ///
    /// Upload a single file and get the corresponding URI.  &#x60;POST {{ api_url }}/v1/user_uploads&#x60;  Initially, only you will be able to access the link.  To share the uploaded file, you&#39;ll need to [send a message][send-message] containing the resulting link.  Users who can already access the link can reshare it with other users by sending additional Zulip messages containing the link.  [uploaded-files]: /help/manage-your-uploaded-files [send-message]: /api/send-message 
    @PostReq(path: "/user_uploads")
    Future<JsonSuccessBase> uploadFile(
            
            @AsMultipartField() MultipartFile filename
        ) {
        return super.uploadFile(

        
        filename
        ).timeout(timeout);
    }


}

File: scala-play-server/app/api/MessagesApi.scala
package api

import play.api.libs.json._
import model.CodedError
import model.InvalidMessageError
import model.JsObject
import model.JsonSuccess
import model.JsonSuccessBase
import model.OasAnyTypeNotMapped
import model.OneOfobjectobject
import model.OneOfstringinteger
import play.api.libs.Files.TemporaryFile

@javax.annotation.Generated(value = Array("org.openapitools.codegen.languages.ScalaPlayFrameworkServerCodegen"))
trait MessagesApi {
  /**
    * Add an emoji reaction
    * Add an [emoji reaction](/help/emoji-reactions) to a message.  &#x60;POST {{ api_url }}/v1/messages/{message_id}/reactions&#x60; 
    * @param messageId The target message&#39;s ID. 
    * @param emojiName The target emoji&#39;s human-readable name.  To find an emoji&#39;s name, hover over a message to reveal three icons on the right, then click the smiley face icon. Images of available reaction emojis appear. Hover over the emoji you want, and note that emoji&#39;s text name. 
    * @param emojiCode A unique identifier, defining the specific emoji codepoint requested, within the namespace of the &#x60;reaction_type&#x60;.  For most API clients, you won&#39;t need this, but it&#39;s important for Zulip apps to handle rare corner cases when adding/removing votes on an emoji reaction added previously by another user.  If the existing reaction was added when the Zulip server was using a previous version of the emoji data mapping between Unicode codepoints and human-readable names, sending the &#x60;emoji_code&#x60; in the data for the original reaction allows the Zulip server to correctly interpret your upvote as an upvote rather than a reaction with a \&quot;diffenent\&quot; emoji. 
    * @param reactionType If an app is adding/removing a vote on an existing reaction, it should pass this parameter using the value the server provided for the existing reaction for specificity.  Supported values:  * &#x60;unicode_emoji&#x60;: Unicode emoji (&#x60;emoji_code&#x60; will be its Unicode codepoint). * &#x60;realm_emoji&#x60;: Custom emoji. (&#x60;emoji_code&#x60; will be its ID). * &#x60;zulip_extra_emoji&#x60;: Special emoji included with Zulip.  Exists to    namespace the &#x60;zulip&#x60; emoji.  **Changes**: In Zulip 3.0 (feature level 2), this become optional for [custom emoji](/help/add-custom-emoji); previously, this endpoint assumed &#x60;unicode_emoji&#x60; if this parameter was not specified. 
    */
  def addReaction(messageId: Int, emojiName: String, emojiCode: Option[String], reactionType: Option[String]): JsonSuccess

  /**
    * Check if messages match a narrow
    * Check whether a set of messages match a [narrow](/api/construct-narrow).  &#x60;GET {{ api_url }}/v1/messages/matches_narrow&#x60;  For many common narrows (E.g. a topic), clients can write an efficient client-side check to determine whether a newly arrived message belongs in the view.  This endpoint is designed to allow clients to handle more complex narrows for which the client does not (or in the case of full-text search, cannot) implement this check.  The format of the &#x60;match_subject&#x60; and &#x60;match_content&#x60; objects is designed to match those of &#x60;GET /messages&#x60;, so that a client can splice these fields into a &#x60;message&#x60; object received from &#x60;GET /events&#x60; and end up with an extended message object identical to how a &#x60;GET /messages&#x60; for the current narrow would have returned the message. 
    * @param msgIds List of IDs for the messages to check.
    * @param narrow A structure defining the narrow to check against. See how to [construct a narrow](/api/construct-narrow).
    */
  def checkMessagesMatchNarrow(msgIds: List[Int], narrow: List[JsObject]): OasAnyTypeNotMapped

  /**
    * Delete a message
    * Permanently delete a message.  &#x60;DELETE {{ api_url }}/v1/messages/{msg_id}&#x60;  This API corresponds to the [delete a message completely][delete-completely] feature documented in the Zulip Help Center.  [delete-completely]: /help/edit-or-delete-a-message#delete-a-message-completely 
    * @param messageId The target message&#39;s ID. 
    */
  def deleteMessage(messageId: Int): JsonSuccess

  /**
    * Get public temporary URL
    * Get a temporary URL for access to the file that doesn&#39;t require authentication. 
    * @param realmIdStr The realm id. 
    * @param filename Path to the URL. 
    */
  def getFileTemporaryUrl(realmIdStr: Int, filename: String): JsonSuccessBase

  /**
    * Get a message&#39;s edit history
    * Fetch the message edit history of a previously edited message.  &#x60;GET {{ api_url }}/v1/messages/{message_id}/history&#x60;  Note that edit history may be disabled in some organizations; see the [Zulip Help Center documentation on editing messages][edit-settings].  [edit-settings]: /help/view-a-messages-edit-history 
    * @param messageId The target message&#39;s ID. 
    */
  def getMessageHistory(messageId: Int): JsonSuccessBase

  /**
    * Get messages
    * Fetch message history from a Zulip server.  &#x60;GET {{ api_url }}/v1/messages&#x60;  This &#x60;GET /api/v1/messages&#x60; endpoint is the primary way to fetch message history from a Zulip server.  It is useful both for Zulip clients (e.g. the web, desktop, mobile, and terminal clients) as well as bots, API clients, backup scripts, etc.  By specifying a [narrow filter](/api/construct-narrow), you can use this endpoint to fetch the messages matching any search query that is supported by Zulip&#39;s powerful full-text search backend.  When a narrow is not specified, it can be used to fetch a user&#39;s message history. (We recommend paginating to 1000 messages at a time.)  In either case, you specify an &#x60;anchor&#x60; message (or ask the server to calculate the first unread message for you and use that as the anchor), as well as a number of messages before and after the anchor message.  The server returns those messages, sorted by message ID, as well as some metadata that makes it easy for a client to determine whether there are more messages matching the query that were not returned due to the &#x60;num_before&#x60; and &#x60;num_after&#x60; limits.  We recommend using &#x60;num_before &lt;&#x3D; 1000&#x60; and &#x60;num_after &lt;&#x3D; 1000&#x60; to avoid generating very large HTTP responses. A maximum of 5000 messages can be obtained per request; attempting to exceed this will result in an error. 
    * @param numBefore The number of messages with IDs less than the anchor to retrieve. 
    * @param numAfter The number of messages with IDs greater than the anchor to retrieve. 
    * @param anchor Integer message ID to anchor fetching of new messages. Supports special string values for when the client wants the server to compute the anchor to use:  * &#x60;newest&#x60;: The most recent message. * &#x60;oldest&#x60;: The oldest message. * &#x60;first_unread&#x60;: The oldest unread message matching the   query, if any; otherwise, the most recent message.  The special values of &#x60;&#39;newest&#39;&#x60; and &#x60;&#39;oldest&#39;&#x60; are also supported for anchoring the query at the most recent or oldest messages.  **Changes**: String values are new in Zulip 3.0 (feature level 1).  The   &#x60;first_unread&#x60; functionality was supported in Zulip 2.1.x   and older by not sending anchor and using use_first_unread_anchor.    In Zulip 2.1.x and older, &#x60;oldest&#x60; can be emulated with   &#x60;anchor&#x3D;0&#x60;, and &#x60;newest&#x60; with &#x60;anchor&#x3D;10000000000000000&#x60;   (that specific large value works around a bug in Zulip   2.1.x and older in the &#x60;found_newest&#x60; return value). 
    * @param narrow The narrow where you want to fetch the messages from. See how to [construct a narrow](/api/construct-narrow). 
    * @param clientGravatar Whether the client supports computing gravatars URLs.  If enabled, &#x60;avatar_url&#x60; will be included in the response only if there is a Zulip avatar, and will be &#x60;null&#x60; for users who are using gravatar as their avatar.  This option significantly reduces the compressed size of user data, since gravatar URLs are long, random strings and thus do not compress well. The &#x60;client_gravatar&#x60; field is set to &#x60;true&#x60; if clients can compute their own gravatars. 
    * @param applyMarkdown If &#x60;true&#x60;, message content is returned in the rendered HTML format. If &#x60;false&#x60;, message content is returned in the raw Markdown-format text that user entered. 
    * @param useFirstUnreadAnchor Legacy way to specify &#x60;anchor&#x3D;\&quot;first_unread\&quot;&#x60; in Zulip 2.1.x and older.  Whether to use the (computed by the server) first unread message matching the narrow as the &#x60;anchor&#x60;.  Mutually exclusive with &#x60;anchor&#x60;.  **Changes**: Deprecated in Zulip 3.0, replaced by &#x60;anchor&#x3D;\&quot;first_unread\&quot;&#x60; instead. 
    */
  def getMessages(numBefore: Int, numAfter: Int, anchor: Option[OneOfstringinteger], narrow: Option[List[JsObject]], clientGravatar: Option[Boolean], applyMarkdown: Option[Boolean], useFirstUnreadAnchor: Option[Boolean]): JsonSuccessBase

  /**
    * Get a message&#39;s raw Markdown
    * Get the raw content of a message.  &#x60;GET {{ api_url }}/v1/messages/{msg_id}&#x60;  This is a rarely-used endpoint relevant for clients that primarily work with HTML-rendered messages but might need to occasionally fetch the message&#39;s raw Markdown (e.g. for pre-filling a message-editing UI). 
    * @param messageId The target message&#39;s ID. 
    */
  def getRawMessage(messageId: Int): JsonSuccessBase

  /**
    * Mark all messages as read
    * Marks all of the current user&#39;s unread messages as read.  &#x60;POST {{ api_url }}/v1/mark_all_as_read&#x60; 
    */
  def markAllAsRead(): JsonSuccess

  /**
    * Mark messages in a stream as read
    * Mark all the unread messages in a stream as read. 
    * @param streamId The ID of the stream to access. 
    */
  def markStreamAsRead(streamId: Int): JsonSuccess

  /**
    * Mark messages in a topic as read
    * Mark all the unread messages in a topic as read. 
    * @param streamId The ID of the stream to access. 
    * @param topicName The name of the topic whose messages should be marked as read. 
    */
  def markTopicAsRead(streamId: Int, topicName: String): JsonSuccess

  /**
    * Remove an emoji reaction
    * Remove an [emoji reaction](/help/emoji-reactions) from a message.  &#x60;DELETE {{ api_url }}/v1/messages/{message_id}/reactions&#x60; 
    * @param messageId The target message&#39;s ID. 
    * @param emojiName The target emoji&#39;s human-readable name.  To find an emoji&#39;s name, hover over a message to reveal three icons on the right, then click the smiley face icon. Images of available reaction emojis appear. Hover over the emoji you want, and note that emoji&#39;s text name. 
    * @param emojiCode A unique identifier, defining the specific emoji codepoint requested, within the namespace of the &#x60;reaction_type&#x60;.  For most API clients, you won&#39;t need this, but it&#39;s important for Zulip apps to handle rare corner cases when adding/removing votes on an emoji reaction added previously by another user.  If the existing reaction was added when the Zulip server was using a previous version of the emoji data mapping between Unicode codepoints and human-readable names, sending the &#x60;emoji_code&#x60; in the data for the original reaction allows the Zulip server to correctly interpret your upvote as an upvote rather than a reaction with a \&quot;diffenent\&quot; emoji. 
    * @param reactionType If an app is adding/removing a vote on an existing reaction, it should pass this parameter using the value the server provided for the existing reaction for specificity.  Supported values:  * &#x60;unicode_emoji&#x60;: Unicode emoji (&#x60;emoji_code&#x60; will be its Unicode codepoint). * &#x60;realm_emoji&#x60;: Custom emoji. (&#x60;emoji_code&#x60; will be its ID). * &#x60;zulip_extra_emoji&#x60;: Special emoji included with Zulip.  Exists to    namespace the &#x60;zulip&#x60; emoji.  **Changes**: In Zulip 3.0 (feature level 2), this become optional for [custom emoji](/help/add-custom-emoji); previously, this endpoint assumed &#x60;unicode_emoji&#x60; if this parameter was not specified. 
    */
  def removeReaction(messageId: Int, emojiName: Option[String], emojiCode: Option[String], reactionType: Option[String]): JsonSuccess

  /**
    * Render message
    * Render a message to HTML.  &#x60;POST {{ api_url }}/v1/messages/render&#x60; 
    * @param content The content of the message. Maximum message size of 10000 bytes. 
    */
  def renderMessage(content: String): JsonSuccessBase

  /**
    * Send a message
    * Send a stream or a private message.  &#x60;POST {{ api_url }}/v1/messages&#x60; 
    * @param `type` The type of message to be sent. &#x60;private&#x60; for a private message and &#x60;stream&#x60; for a stream message. 
    * @param to For stream messages, either the name or integer ID of the stream. For private messages, either a list containing integer user IDs or a list containing string email addresses.  **Changes**: Support for using user/stream IDs was added in Zulip 2.0.0. 
    * @param content The content of the message. Maximum message size of 10000 bytes. 
    * @param topic The topic of the message. Only required for stream messages (&#x60;type&#x3D;\&quot;stream\&quot;&#x60;), ignored otherwise.  Maximum length of 60 characters.  **Changes**: New in Zulip 2.0.  Previous Zulip releases encoded this as &#x60;subject&#x60;, which is currently a deprecated alias. 
    * @param queueId For clients supporting [local echo](https://zulip.readthedocs.io/en/latest/subsystems/sending-messages.html#local-echo), the [event queue](/api/register-queue) ID for the client.  If passed, &#x60;local_id&#x60; is required.  If the message is successfully sent, the server will include &#x60;local_id&#x60; in the &#x60;message&#x60; event that the client with this &#x60;queue_id&#x60; will receive notifying it of the new message via [&#x60;GET /events&#x60;](/api/get-events).  This lets the client know unambiguously that it should replace the locally echoed message, rather than adding this new message (which would be correct if the user had sent the new message from another device). 
    * @param localId For clients supporting local echo, a unique string-format identifier chosen freely by the client; the server will pass it back to the client without inspecting it, as described in the &#x60;queue_id&#x60; description. 
    */
  def sendMessage(`type`: String, to: List[Int], content: String, topic: Option[String], queueId: Option[String], localId: Option[String]): JsonSuccessBase

  /**
    * Edit a message
    * Edit/update the content or topic of a message.  &#x60;PATCH {{ api_url }}/v1/messages/{msg_id}&#x60;  &#x60;{msg_id}&#x60; in the above URL should be replaced with the ID of the message you wish you update.  You can [resolve topics](/help/resolve-a-topic) by editing the topic to &#x60;✔ {original_topic}&#x60;.  **Note**: See [configuring message editing][config-message-editing] for detailed documentation on when users are allowed to edit topics.  [config-message-editing]: /help/configure-message-editing-and-deletion 
    * @param messageId The target message&#39;s ID. 
    * @param topic The topic to move the message(s) to, to request changing the topic. Should only be sent when changing the topic, and will throw an error if the target message is not a stream message.  Maximum length of 60 characters.  **Changes**: New in Zulip 2.0.  Previous Zulip releases encoded this as &#x60;subject&#x60;, which is currently a deprecated alias. 
    * @param propagateMode Which message(s) should be edited: just the one indicated in &#x60;message_id&#x60;, messages in the same topic that had been sent after this one, or all of them.  Only the default value of &#x60;change_one&#x60; is valid when editing only the content of a message.  This parameter determines both which messages get moved and also whether clients that are currently narrowed to the topic containing the message should navigate or adjust their compose box recipient to point to the post-edit stream/topic. 
    * @param sendNotificationToOldThread Whether to send breadcrumb message to the old thread to notify users where the messages were moved to.  **Changes**: New in Zulip 3.0 (feature level 9). 
    * @param sendNotificationToNewThread Whether to send a notification message to the new thread to notify users where the messages came from.  **Changes**: New in Zulip 3.0 (feature level 9). 
    * @param content The content of the message. Maximum message size of 10000 bytes. 
    * @param streamId The stream ID to move the message(s) to, to request moving messages to another stream.  Should only be sent when changing the stream, and will throw an error if the target message is not a stream message. 
    */
  def updateMessage(messageId: Int, topic: Option[String], propagateMode: Option[String], sendNotificationToOldThread: Option[Boolean], sendNotificationToNewThread: Option[Boolean], content: Option[String], streamId: Option[Int]): JsonSuccess

  /**
    * Update personal message flags
    * Add or remove personal message flags like &#x60;read&#x60; and &#x60;starred&#x60; on a collection of message IDs.  &#x60;POST {{ api_url }}/v1/messages/flags&#x60;  For updating the &#x60;read&#x60; flag on common collections of messages, see also the [special endpoints for marking message as read in bulk](/api/mark-all-as-read). 
    * @param messages An array containing the IDs of the target messages. 
    * @param op Whether to &#x60;add&#x60; the flag or &#x60;remove&#x60; it. 
    * @param flag The flag that should be added/removed. 
    */
  def updateMessageFlags(messages: List[Int], op: String, flag: String): JsonSuccessBase

  /**
    * Upload a file
    * Upload a single file and get the corresponding URI.  &#x60;POST {{ api_url }}/v1/user_uploads&#x60;  Initially, only you will be able to access the link.  To share the uploaded file, you&#39;ll need to [send a message][send-message] containing the resulting link.  Users who can already access the link can reshare it with other users by sending additional Zulip messages containing the link.  [uploaded-files]: /help/manage-your-uploaded-files [send-message]: /api/send-message 
    */
  def uploadFile(filename: Option[TemporaryFile]): JsonSuccessBase
}
